### DISCUSSION TITLE
How to retrain the Base-RCNN-FPN with custom dataset?

### DISCUSSION THREAD

**Discussion by divyadey7** [facebookresearch/detectron2#4485] [Q&A]
I want to retrain the base detection model used for Keypoint-RCNN. How can I do it with my custom dataset of person class? Is there a way to add the custom dataset with annotations to the coco_2017 training set and retrain the model? Or if there's any other way of doing it?

**Comment by ExtReMLapin** [comment]
You mean transfer learning ?

Here is what I do, I get all my custom classes, let's say i got the following:

* 900 instances of `rifle`
* 500 instances of `grenade`
* 730 instances of `pistol`

Now this is my custom dataset, I might have in this some existing classes of the coco dataset.

In all my dataset, even if I'm supposed to have some kind of evening/equilibration between the number of instance of each classes, i'll take the class with the most instances. `900` here.


I'll take my whole dataset, take 20% out of it randomly to create my test set.

Take the coco dataset (train and validate)
Build an index of it, it might take some space, inside try to get this kind of data, so you don't need to search later for the class you want to add to your training/test dataset.
![image](https://user-images.githubusercontent.com/3909752/186847904-d96d2853-7081-4001-b764-fc50cf353954.png)
![image](https://user-images.githubusercontent.com/3909752/186848010-3011fdc2-034b-4185-88bc-88d228696bb3.png)
![image](https://user-images.githubusercontent.com/3909752/186848067-169db962-fc40-40e9-a0fa-b285788ff0b1.png)


For each of the instance of the coco dataset, get at LEAST `900` instances of each classes , take 20% and merge it with your test set
Take the 80% left and merge it with your trainning set.

With this, it allows you to keep the "knowledge" of your old coco classes while learning new custom classes.

I did this with canabis flowers, to prevent mixing up broccoli and cannabis flower and it works great

**Reply by divyadey7** [reply]
will it follow the same if I am about to retrain the existing classes? like I want to retrain the Base Detection model for person class and use this model for keypoint detection?

**Reply by ExtReMLapin** [reply]
Isn't that two separate models ?

**Reply by opyate** [reply]
> With this, it allows you to keep the "knowledge" of your old coco classes while learning new custom classes.

That's an interesting idea. Did you stumble upon this empirically, or is it based on an established technique?

**Reply by ExtReMLapin** [reply]
It was just a guess, we gave a try and it worked, just used grey matter, I wouldn't be surprised someone else did it before
